First draft of 1DRO unit model

adam-a-a opened the pull request:
## Fixes/Addresses: #126 

## Summary/Motivation:

This is a first attempt at developing the 1D-RO model. While more detail will be added subsequently, a simple version of the 1D-RO model is presented here.

## Changes proposed in this PR:
- 1D control volume on feed side
- Indexed state block for permeate side
- State block for the final permeate exiting the module
- initial testing of the model
- initial technical spec document as a placeholder (some simple info was filled in, but will need more work in future PR)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

adam-a-a commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,824 @@
+            doc="""Discretization scheme to use when transforming domain. See
+    Pyomo documentation for supported schemes."""))
+
+
```
```suggestion
```

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+    **MomentumBalanceType.momentumTotal** - single momentum balance for material,
+    **MomentumBalanceType.momentumPhase** - momentum balances for each phase.}"""))
+
+    CONFIG.declare("finite_elements", ConfigValue(
```
Just as a matter of organization, I would keep all the DAE related arguments together. At the moment, the default balance type arguments are stuck in the middle of them.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+
+        self.scaling_factor = Suffix(direction=Suffix.EXPORT)
+
+        self.solvent_list = Set()
```
You should be able to get these from the property package, e.g. `self.config.property_package.solvent_set`

adam-a-a replied:
I think that `solvent_set` and `solute_set` weren't defined in the NaCl property model, so would have to modify the property model for that to work.

andrewlee94 replied:
That is true. That said, from the IDAES standards perspective at least, it is the property package's job to define these sorts of things.

TimBartholomew replied:
Yes the solvent_set and solute_set should be defined in the property package. It really should be automatically created from defining the components like the component_list and phase_list on IDAES-side. I was planning on making a PR for IDAES, but it got deprioritized.

andrewlee94 replied:
The generic property framework already does this, so it is only the custom property packages which lack these, Making a general PR would need to keep in mind how this would impact the generic framework.

adam-a-a replied:
For now, I will leave this be until we move forward on a PR for the property model. Also want to note that we'd have to revise this segment of code in the 0D RO model as well.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+    **MaterialBalanceType.elementTotal** - use total element balances,
+    **MaterialBalanceType.total** - use total material balance.}"""))
+
+    CONFIG.declare("energy_balance_type", ConfigValue(
```
Later on you set energy balance type to `None`. Either don't add it here as an option, or make use of this argument later (which would require supporting energy balances).

adam-a-a replied:
For now, I will delete the option and add support for energy balances after I get other parts working.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+        # Populate feed side
+        feed_side.add_material_balances(balance_type=self.config.material_balance_type,
+                                        has_mass_transfer=True)
+        feed_side.add_energy_balances(balance_type=EnergyBalanceType.none)
```
As noted above.

However, also note that the underlying bug in the ideal property package was fixed such that you should be able to do an energy balance now and not see excessive temperature change due to a change in pressure.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+        # Add inlet/outlet ports for feed side
+        self.add_inlet_port(name="feed_inlet", block=feed_side)
+        self.add_outlet_port(name="feed_outlet", block=feed_side)
+        #TODO: Make permeate_side.permeate_out a ProductBlock instead?
```
No, you don't want a ProductBlock (although it would work). Feed and Product blocks are just wrappers around a StateBlock anyway, with an appropriately named Port (inlet for a Product block) and some extra code to allow users to do `Product.state_var.fix`.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+
+        # Add reference to pressure drop for feed side only
+        if (self.config.has_pressure_change is True and
+                self.config.momentum_balance_type != 'none'):
```
Need to use the MomentumBlaanceType Enum here.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+
+        # TODO: Add initial value, bounds and real equation later. Just need for now since add_geometry() generates
+        #  cross-sectional area var that I am referencing
+        # self.feed_area_cross = Var(
```
Is this code needed in the released version?

adam-a-a replied:
Not until I add on additional variables and constraints for detailed calculations that account for concentration polarization (which will be in a subsequent PR).

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+            units=units_meta('length'),
+            doc='Membrane width')
+
+        # @self.Constraint(doc="Cross-sectional area of feed channel")
```
Again, commented code.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+        # mass transfer
+        def mass_transfer_phase_comp_initialize(b, t, x, p, j):
+            return value(self.feed_side.properties[t, x].get_material_flow_terms('Liq', j)
+                         * 0.5) # self.recovery_mass_phase_comp[t, 'Liq', j])
```
Trailing comment,

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+            self.feed_side.length_domain,
+            self.config.property_package.phase_list,
+            self.config.property_package.component_list,
+            initialize=0.1,  # mass_transfer_phase_comp_initialize,
```
Trailing comment,

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+                return Constraint.Skip
+            else:
+                return (b.permeate_side[t, x].get_material_flow_terms(p, j)
+                        == -b.feed_side.mass_transfer_term[t, x, p, j] * b.length / nfe)
```
Be careful here - this is only the number of finite elements. If the user selected a collocation transformation there will be more points than this.

`len(self.feed_side.length_domain)-1` might be more useful.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+            domain=NonNegativeReals,
+            units=units_meta('mass') * units_meta('time')**-1 * units_meta('length')**-1,
+            doc='Mass transfer to permeate')
+        # ==========================================================================
```
More commented code.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+            outlvl=outlvl,
+            optarg=optarg,
+            solver=solver,
+            state_args=permeate_side_args)
```
Extension for the future, but is there a way you could estimate some of these based on the feed state? Notably, once you have the permeate side, it is probably possible to directly calculate the permeate outlet state.

adam-a-a replied:
Yes, I definitely would like to improve this. This is my first time writing an initialization routine like this, so still figuring it out things as I go. Would you happen to have any recommendations for resources on developing (getting started with developing) initialization routines? So far, I've just looked at several unit models in IDAES and followed along with the logic.

andrewlee94 replied:
Unfortunately, there are no good resources I am aware of. As far as examples go, the gas-solids contactors have some of the most complex initialisation routines, but they might also not be easy to follow.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+
+    def _get_performance_contents(self, time_point=0):
+        pass
+        # var_dict = {}
```
This should be cleaned up.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+
+    def _get_stream_table_contents(self, time_point=0):
+        pass
+        # return create_stream_table_dataframe(
```
As above.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+###############################################################################
+
+
+from enum import Enum
```
You do not appear to be using this import.

adam-a-a replied:
I was planning to use this when I add in options for concentration polarization effects and pressure drop assumptions. I could delete now but would be bringing it right back in a subsequent PR (or perhaps just build on this one?)

andrewlee94 replied:
It is generally good practice to only include imports you use. You add the import back once you add the new options.

andrewlee94 commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,800 @@
+
+from enum import Enum
+# Import Pyomo libraries
+from pyomo.environ import (Var,
```
Check your imports, I am pretty certain there area few here you don't use.

E.g. I do not recall seeing `Reference` used (you used `add_object_reference` instead.

andrewlee94 reviewed (changes requested):
I haven't looked at the docs or the test yet, but here are some comments on the unit model.

adam-a-a commented:
@TimBartholomew- before I go ahead with enhancements (adding CP, pressure drop), I am contemplating whether it still makes sense to have _all_ of the same options for the 1d model as we do for the 0D RO model.

In terms of consistency, it would be good to apply the same options to the 1D. Additionally, users could take results they may already have from their own calculations (where simplifying assumptions were previously applied) and compare them with results from the 1D model in Proteuslib.

On the other hand, while allowing the user to fix the CP modulus, for example, is meant to facilitate the use of simplifying assumptions, would such simplifications be necessary when the user has chosen to implement a 1-D model? In other words, we've already taken a step up in complexity by shifting to a 1-D model, so does it still make sense to support the same simplifying assumptions as before? What are your thoughts on this?

adam-a-a commented:
@andrewlee94 @TimBartholomew I made most of the requested changes (haven't modified the initialization method just yet). I also add content to the two methods tied to the report() method. Good thing I did this because I then observed that the temperature at the outlet of the feed channel was ~ 360 C when the inlet was set to 298 C. 

I suppose I didn't add any isothermal constraints for temperature along the feed channel but didn't realize I needed to. Also, I excluded the addition of energy balances, so I am guessing that adding that back in might resolve the issue. My only concern is getting the DOF around 0, which I had to fiddle with a bit.

adam-a-a commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,726 @@
+        # Populate feed side
+        feed_side.add_material_balances(balance_type=self.config.material_balance_type,
+                                        has_mass_transfer=True)
+        # feed_side.add_energy_balances(balance_type=self.config.energy_balance_type,
```
```suggestion
```

adam-a-a commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,726 @@
+        feed_side.add_material_balances(balance_type=self.config.material_balance_type,
+                                        has_mass_transfer=True)
+        # feed_side.add_energy_balances(balance_type=self.config.energy_balance_type,
+        #                               has_enthalpy_transfer=True)
```
```suggestion
```

TimBartholomew commented on docs/technical_reference/unit_models/reverse_osmosis_1D.rst:
```diff
@@ -0,0 +1,169 @@
+    Everything below this section is pending since the 1D-RO model is currently under development.
+
+Variables
+----------
```
Instead of duplicating everything, I think it would be best to just refer readers to the 0D RO model and just note the differences between the 0D and 1D model here. That way we only have one place to change the documentation when we make a change in some variable or equation.

adam-a-a replied:
That works for me. This was meant to be more of a placeholder until the 1dRO model was finished. I'd address this in a subsequent PR.

adam-a-a replied:
Actually, just added the links to the relevant 0DRO doc sections until all vars/constraints are settled on (shouldn't be much difference, if any).

TimBartholomew commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,724 @@
+            self.feed_side.length_domain,
+            doc="Material properties of permeate along permeate channel",
+            default=tmp_dict)
+        self.permeate_out = self.config.property_package.state_block_class(
```
We may want to call this permeate_mixed for consistency with the 0D RO model. Although I'm up for changing names.

adam-a-a replied:
Similar to my other response, I think we should have a separate PR (once 1DRO is fully functional) that addresses naming consistency (and things of that nature) between 0D and 1D.

TimBartholomew commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,724 @@
+        tmp_dict["has_phase_equilibrium"] = False
+        tmp_dict["parameters"] = self.config.property_package
+        tmp_dict["defined_state"] = False  # these blocks are not inlets
+        self.permeate_side = self.config.property_package.state_block_class(
```
I think calling these stateblocks "permeate_side" could be confusing because "permeate_side" is a control volume in the 0D model. We should definitely decide whether we want the permeate side to have control volume without mass and energy balances or not and be consistent across the 1D and 0D model.

adam-a-a replied:
During my first attempt to implement the 1DRO model, I tried setting the permeate_side as a 1D control volume, but it made the implementation trickier. Furthermore, after some discussion in issue #126, I decided to pivot and set the permeate_side as an indexed stateblock. I think this is something we can revisit in a subsequent PR, once the 1DRO model is fully functional. The same applies for any name changing or consistency with the 0DRO model; let's reserve such changes for when functionality of the 1DRO in place. Then we can have a PR that cleans up discrepancies between the 0D and 1D models where possible.

andrewlee94 replied:
I personally wouldn't be too concerned. If someone is digging in deep enough to find this, then they had better know enough to understand what is happening anyway.

TimBartholomew commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,724 @@
+
+        self._make_performance()
+
+    def _make_performance(self):
```
We don't have a similar internal function for variable and constraint creation in the 0D RO model, is this something that was implemented in another 1D model or is necessary for 1D models?

adam-a-a replied:
It was something that I came across in some IDAES models and seems to be a matter of preference. I implemented that approach here for my own readability, but we can revert this back to the way vars and constraints are implemented in 0DRO (or modify 0DRO to match this). However, I'd prefer to get the 1DRO model working fully before making such changes.

andrewlee94 replied:
Opinion is somewhat divided in IDAES. Most models don't do it, but they are fairly simple and were all written by me. Others seem to prefer having these sub-methods, and they do have some advantages for testing (if you unit test each method separately, which no-one does).

TimBartholomew commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -0,0 +1,724 @@
+        return {"vars": var_dict}
+
+    def _get_stream_table_contents(self, time_point=0):
+        pass
```
I think you need to remove pass here

adam-a-a replied:
```suggestion
```

TimBartholomew reviewed (commented):
Some comments. I think the main thing is we should make sure that 0D and 1D model are consistent, we may plan to change the 0D model to match what is here.

adam-a-a commented:
@TimBartholomew @andrewlee94 I think I addressed all of your comments. With regard to naming consistency and similar issues between 0DRO and 1DRO, that would be settled in a later PR. Right now, I'd like to get this PR through since I have another PR for 1DRO ready to go. Do you have any other  other comments or suggested changes?

TimBartholomew reviewed (approved):
This looks fine and we can address the consistency across 0D and 1D RO when we decide on how we are refactoring the 0D model.
